### DISCUSSION TITLE
Optimization hints

### DISCUSSION THREAD

**Discussion by KarolS** [KarolS/millfork#93] [Ideas]
There's work undergoing to implement optimization hints.

The goal is to provide annotations that can help the compiler to generate faster code. The underlying philosophy is:

* it is always correct to ignore optimization hints, at least assuming timing is not an issue
* as a consequence, unknown hints should be a compilation warning, not an error
* applying incorrectly specified hints may break the correctness of the code, it's up to the programmer to specify the hints correctly

Here is what is done:

* marking functions as preserving registers
* marking functions as not modifying memory
* marking functions as hot
* marking functions as idempotent
* marking functions as returning even/odd values

Here's what remains to be done:

* syntax (I'm using `¥(...)` as a placeholder for testing)
* marking functions as cold
* marking functions as requiring a specific DP value (useful for implementing a fast and convenient Vectrex library)
* marking variables (with tons of various hints, ideas are welcome)
* marking function parameters (as they're essentially variables)

Examples:

* CBM CHROUT routine actually preserves all registers, so it should be specified as `void putchar(byte register(a) char) ¥(preserves_a, preserves_x, preserves_y) @$ffd2 extern`, letting the code that calls it to use the registers optimally

**Comment by agg23** [comment]
> marking variables (with tons of various hints, ideas are welcome)

I'm really interested to see what optimizations can come out of this. Off of the top of my head, it would be really nice to allow for local variable reuse.
